### PR TITLE
[Frontend] Chat-based Embeddings API

### LINKED ISSUE #8967: [RFC]: Reward Modelling in OpenAI Compatible Server
### Motivation.

Reward models are an important tool in NLP / AI workflows, especially in agentic flows which use them to verify quality of intermediate outputs, or rank between several attempts at performing a single task.

vLLM just added support for a reward model in https://github.com/vllm-project/vllm/pull/8896#issuecomment-2381912003 .
This requires a workaround in order to work with the OpenAI Compatible Server - it piggybacks the embedding endpoint.
The workaround requires the client to know which tokenizer is being used by the server, apply the chat template to the conversation, and send the resulting string to the embedding endpoint. This isn't ideal and it breaks the decoupling between the client and server.

A short discussion in the same issue led to the creation of the RFC.

### Proposed Change.

The reason that no endpoint currently matches the needs of the reward model is as follows:
- The embedding endpoint receives a string as the input, not a conversation
- The chat endpoint returns a string, not a series of numbers. Even if you ask for logprobs, they are after softmax was applied, which is not a reversable process.

I see several ways to more elegantly support reward models in the OpenAI compatible server, and this RFC will hopefully be the discussion point for them.

Option 1:
Add a conversation object (`List[Dict[str, str]]`) as a potential input to `EmbeddingRequest` class. It already supports a variety of options:
``` input: Union[List[int], List[List[int]], str, List[str]]```
Upon detecting that a conversation object was given, the OpenAI Compatible server will apply the chat template using the tokenizer, and proceed as if it received `str` input.

Option 2:
Add a way to get output logits instead of output logprobs from the chat endpoint. This can be either a new per-request parameter (similar to `top_logprobs`) or a server-side flag to override the behavior of the data returned in the field (`--return_logits_instead_of_logprobs` flag to the OpenAI server for example).

Option 3:
Add a dedicated endpoint to vLLM.

Option 4:
Do nothing. Since there is a `/tokenize` endpoint that also accepts a conversation, the sample code in https://github.com/vllm-project/vllm/pull/8896#issuecomment-2381912003 could be changed to use the `tokenize` endpoint, receieve the tokens list and send that the `embeddings` endpoint, which addresses the coupling problem.

I personally support Option 1, as it feels the least hacky of the bunch, and also does not require a whole lot of new code.

What do you think?

### Feedback Period.

Not my decision to say when a conclusion was reached here, but I don't think it should take more than a couple of weeks.

### CC List.

@simon-mo 
@DarkLight1337 
@zhuzilin 

### Any Other Things.

vLLM is awesome!

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR extends the existing Embeddings API to accept chat conversations similar to Chat Completions API. This enables multi-modal conversations to be passed to the embedding model.

To reduce code duplication, I've also factored out the common code for handling completion and chat-based inputs into the base `OpenAIServing` class.

FIX #8967
FIX https://github.com/vllm-project/vllm/pull/9303#issuecomment-2437569903

### CODE REVIEW

File: vllm/entrypoints/openai/serving_embedding.py [hunk 1]
```diff
@@ -113,15 +133,6 @@ async def create_embedding(
 
         truncate_prompt_tokens = None
 
-        if request.truncate_prompt_tokens is not None:
```
Comment (maxdebayser) [inline vllm/entrypoints/openai/serving_embedding.py:116]: This verification was added to reproduce the behavior of the OpenAI embeddings API. It returns an error if the sequence length exceeds the model max length.

Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_embedding.py:116]: Oops, didn't mean to remove this!

**Comment by DarkLight1337** [comment]
Sorry for the commit spam, I'm done refining the docs now.

**Comment by DarkLight1337** [comment]
Entrypoints tests pass now.

**Review by Isotr0py — APPROVED** [review]
Overall LGTM since entrypoint tests all passed. See if @ywang96 is OK with the frontend refactor as well.

**Comment by DarkLight1337** [comment]
@simon-mo @njhill do you have time to take a look at this?

File: docs/source/models/vlm.rst [hunk 2]
```diff
@@ -251,5 +254,50 @@ A full code example can be found in `examples/openai_api_client_for_multimodal.p
...
+
+    vllm serve TIGER-Lab/VLM2Vec-Full --task embedding \
+      --trust-remote-code --max-model-len 4096
+
+.. important::
+
+    Since VLM2Vec has the same model architecture as Phi-3.5-Vision, we have to explicitly pass ``--task embedding``
+    to run this model in embedding mode instead of text generation mode.
+
+Since this schema is not defined by OpenAI client, we post a request to the server using the lower-level ``requests`` library:
```
Comment (maxdebayser) [inline docs/source/models/vlm.rst:278]: Just leaving this as a thought here: should we perhaps have a fork of the openai client that support our extensions explicitly?

File: tests/entrypoints/openai/test_embedding.py [hunk 3]
```diff
@@ -108,22 +109,63 @@ async def test_batch_embedding(embedding_client: openai.AsyncOpenAI,
...
+    tokens = tokenizer.encode(prompt, add_special_tokens=False)
+
+    embeddings = response.json()
+    assert embeddings["id"] is not None
+    assert len(embeddings["data"]) == 1
+    assert len(embeddings["data"][0]["embedding"]) == 4096
+    assert embeddings["usage"]["completion_tokens"] == 0
+    assert embeddings["usage"]["prompt_tokens"] == len(tokens)
+    assert embeddings["usage"]["total_tokens"] == len(tokens)
+
```
Comment (maxdebayser) [inline tests/entrypoints/openai/test_embedding.py:151]: Should we perhaps call the non-chat embedding endpoint here with `prompt` and compare the results to make sure that they return the same thing?

File: vllm/pooling_params.py [hunk 4]
```diff
@@ -7,7 +7,7 @@ class PoolingParams(
         msgspec.Struct,
         omit_defaults=True,  # type: ignore[call-arg]
         array_like=True):  # type: ignore[call-arg]
-    """Pooling parameters for pooling.
+    """Pooling parameters for embeddings API.
```
Comment (maxdebayser) [inline vllm/pooling_params.py:10]: I might be missing something, but the `additional_data` attribute doesn't seem to be used anywhere. Which is good, because it can by anything and is passed without validation from the request to the `Pooler.forward()` method as part of the `PoolingMetadata` object. If there is no use case for this, can we remove it in this PR?

**Review by maxdebayser — APPROVED** [review]
I left a few comments but overall this looks good to me.

File: tests/entrypoints/openai/test_embedding.py (hunk 3 above)
Comment (ywang96) [reply tests/entrypoints/openai/test_embedding.py:151]: +1

File: vllm/entrypoints/openai/protocol.py [hunk 5]
```diff
@@ -799,7 +880,7 @@ class EmbeddingResponseData(OpenAIBaseModel):
 
 
 class EmbeddingResponse(OpenAIBaseModel):
-    id: str = Field(default_factory=lambda: f"cmpl-{random_uuid()}")
+    id: str = Field(default_factory=lambda: f"embd-{random_uuid()}")
```
Comment (ywang96) [inline vllm/entrypoints/openai/protocol.py:883]: Good catch!

File: vllm/entrypoints/openai/serving_embedding.py [hunk 6]
```diff
@@ -21,7 +23,16 @@
 
 logger = init_logger(__name__)
 
-TypeTokenIDs = List[int]
+
+def check_embedding_mode(model_config: ModelConfig) -> bool:
+    embedding_mode = model_config.task == "embedding"
+
+    if not embedding_mode:
+        logger.warning("embedding_mode is False. Embedding API will not work.")
```
Comment (ywang96) [inline vllm/entrypoints/openai/serving_embedding.py:31]: Shouldn't this be an error? 

or are you thinking an initialized model server can handle both completions and embedding generation? In that case, we should make `model_config.task` a list and check if `embedding` is in it, right?

**Review by ywang96 — COMMENTED** [review]
Left a few comments - PTAL!

File: docs/source/models/vlm.rst [hunk 7]
```diff
@@ -243,6 +243,9 @@ To consume the server, you can use the OpenAI client like in the example below:
 
 A full code example can be found in `examples/openai_api_client_for_multimodal.py <https://github.com/vllm-project/vllm/blob/main/examples/openai_api_client_for_multimodal.py>`_.
 
+.. tip::
+    There is no need to format the prompt in the API request since it will be handled by the server.
+
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply docs/source/models/vlm.rst]: Updated, see if this is clearer now.

File: docs/source/models/vlm.rst (hunk 2 above)
Comment (DarkLight1337) [reply docs/source/models/vlm.rst:279]: This sounds good, but not sure whether we have bandwidth to maintain it 😅

File: tests/entrypoints/openai/test_embedding.py (hunk 3 above)
Comment (DarkLight1337) [reply tests/entrypoints/openai/test_embedding.py:151]: Sounds good!

File: vllm/pooling_params.py (hunk 4 above)
Comment (DarkLight1337) [reply vllm/pooling_params.py:10]: @robertgshaw2-neuralmagic originally added this (https://github.com/vllm-project/vllm/pull/4800#issuecomment-2119278945). I am not sure whether this is still relevant since we can now set the pooling strategy via CLI.

Comment (DarkLight1337) [reply vllm/pooling_params.py:10]: @robertgshaw2-neuralmagic can you comment on this?

**Review by ywang96 — APPROVED** [review]
LGTM!

File: docs/source/models/vlm.rst (hunk 2 above)
Comment (DarkLight1337) [reply docs/source/models/vlm.rst:279]: I suggest opening an issue for this.

File: vllm/pooling_params.py (hunk 4 above)
Comment (DarkLight1337) [reply vllm/pooling_params.py:10]: Meanwhile let's merge this PR first.

File: vllm/entrypoints/openai/serving_completion.py [hunk 8]
```diff
@@ -195,24 +180,30 @@ async def create_completion(
...
         try:
             async for i, res in result_generator:
                 final_res_batch[i] = res
+        except asyncio.CancelledError:
+            return self.create_error_response("Client disconnected")
+        except ValueError as e:
+            # TODO: Use a vllm-specific Validation Error
+            return self.create_error_response(str(e))
 
+        try:
```
Comment (njhill) [inline vllm/entrypoints/openai/serving_completion.py:198]: @DarkLight1337 apologies I didn't get a chance to review this last week .. ran into some things while trying to resolve conflicts with another pending PR :)

It looks like there are a few changes not directly related to chat embeddings.

Wondering the reason for splitting into two try/except blocks here in particular (I guess similar in serving_embedding.py)

Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_completion.py:198]: It is just to make clear that `asyncio.CancelledError` can only happen while iterating through the result generator.
